docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst

```diff
@@ -0,0 +1,451 @@
+-------------------------------------------------------
+
+This example is a bit more advanced, involving both a callback function and a class constructor.
+The Django web framework allows you to specify a function that determines the path where uploaded files are stored (see the `Django documentation <https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.FileField.upload_to>`_).
+This function is passed as an argument to the **FileField** constructor.
+The function is called with two arguments: the instance of the model and the filename of the uploaded file.
+This filename is what we want to mark as a taint source. An example use looks as follows:
+
+.. code-block:: python
+
+  from django.db import models
+
+  def user_directory_path(instance, filename): # <-- add 'filename' as a taint source
+    # file will be uploaded to MEDIA_ROOT/user_<id>/<filename>
+    return "user_{0}/{1}".format(instance.user.id, filename)
+
+  class MyModel(models.Model):
+    upload = models.FileField(upload_to=user_directory_path) # <-- the 'upload_to' parameter defines our custom function
+
+Note that this source is already known by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: sourceModel
+      data:
+        - [
+            "django.db.models.FileField!",
+            "Call.Argument[0,upload_to:].Parameter[1,filenam:]",
+            "remote",
+          ]
+
+
+- Since we're adding a new taint source, we add a tuple to the **sourceModel** extensible predicate.
+- The first column, **"django.db.models.FileField!"**, is a dotted path to the **FileField** class from the **django.db.models** package.
+  The **!** at the end of the type name indicates that we are looking for the class itself rather than instances of this class.
+
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+  
+    - **Call** selects calls to the class. That is, constructor calls.
+    - **Argument[0,upload_to:]** selects the first positional argument, or the named argument named **upload_to**. Note that the colon at the end of the argument name indicates that we are looking for a named argument.
+    - **Parameter[1,filename:]** selects parameters of the callback function. We provide two parameters to search for, namely the second positional parameter and the named parameter **filename**.
+
+- Finally, the kind **remote** indicates that this is considered a source of remote flow.
+
+Example: Adding flow through 're.compile'
+----------------------------------------------
+
+In this example, we'll show how to add flow through calls to `re.compile`.
+`re.compile` returns a compiled regular expression for efficient evaluation, but the pattern to be compiled is stored in the `pattern` attribute of the resulting object.
```

Comment:
While reading through the rendered version, it became clear to me that reStructured text requires double backticks for code snippets ([docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html)).

```suggestion
In this example, we'll show how to add flow through calls to ``re.compile``.
``re.compile`` returns a compiled regular expression for efficient evaluation, but the pattern to be compiled is stored in the ``pattern`` attribute of the resulting object.
```

Reply:
Thanks, it should be replaced everywhere now.

Reply:
Not quite what I asked for.

If you look at the rendered version ([before](https://github.com/github/codeql/blob/4626e134faffd6985dc71c0b7e3af3e24eaaa4ea/docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst), [after](https://github.com/github/codeql/blob/259dbf415305fc3b9403ea4e7f1c2f12842c6d4f/docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst)) the change looks like:

## Before

![image](https://github.com/user-attachments/assets/55cf7ee1-f04b-4bea-8fd7-2560be3ac718)

## After

![image](https://github.com/user-attachments/assets/7e2fe2d3-464b-4ea6-ab98-1649a710b97a)


However, both of them still only use single backticks, meaning the code snippets are not rendered properly.

![image](https://github.com/user-attachments/assets/baae948b-f403-4921-8909-1d48ba0ca5a9)

---

The previous version looked better to me, so I would recommend reverting cd7031a298614cdac95e7dd34fa0550dbb4b8ae7 and 9d62c219c9d2d986a579ad4222f048278cd785d1 (the double-qoutes in the examples made them easier to read for me)

Reply:
> Not quite what I asked for.

Indeed, I realised once I started looking at the generated artifact 🤦 I should have commented then, so you did not have to pull it out. Thanks for doing that, though. I will get the double back-tics in, and revert to single quotes in headers to align with the other documentation pages (e.g. [the one for javascript](https://codeql.github.com/docs/codeql-language-guides/customizing-library-models-for-javascript/))). About the double quotes, I removed them on request, but that was mainly a request for consistency, so perhaps they should be in in a consistent way...

Reply:
> I will get the double back-tics in, and revert to single quotes in headers

:+1:

> About the double quotes, I removed them on request, but that was mainly a request for consistency, so perhaps they should be in in a consistent way...

Ah, I see they are present in the [JS docs](https://codeql.github.com/docs/codeql-language-guides/customizing-library-models-for-javascript/) as well, so :shrug: